### PR TITLE
Initial typesafety testing

### PR DESCRIPTION
**Pull request by booshington**

### CODE REVIEW

File: gremlinapi/config.py [hunk 1]
```diff
@@ -151,11 +151,11 @@ def user_mfa_token_value(self, user_mfa_token_value):
         self._user_mfa_token_value = user_mfa_token_value
         return self.user_mfa_token_value
 
-    @classmethod
-    def is_bearer_expired(cls) -> bool:
```
Comment (booshington) [inline gremlinapi/config.py:155]: @khultman Could we have a sync on this method?

File: tests/test_attacks.py [hunk 2]
```diff
@@ -5,21 +5,46 @@
...
+        test_endpoint = "test-endpoint.com"        
+        expected_output = "%s/?source=scenario&pageSize=3&" % test_endpoint
+        test_kwargs = {"source":"scenario","pageSize":3}
+        test_output = GremlinAPIAttacks._list_endpoint(test_endpoint,**test_kwargs)
+        self.assertEqual(test_output, expected_output)
 
-        self.assertEqual(GremlinAPIAttacks.list_attacks(),mock_data)
\ No newline at end of file
+    def test_error_if_not_attack_body(self) -> None:
+        # expected_output = GremlinAttackHelper(target=GremlinTargetContainers(strategy_type='Random', labels={'owner': 'kyle'}, percent=100),command=GremlinLatencyAttack(delay=1))
```
Comment (booshington) [inline tests/test_attacks.py:46]: @khultman This code seems to hit the API yet I followed it through and I'm not sure exactly how.

**Comment by booshington** [comment]
@khultman All file statically types and covered in unit tests from a function level. The next step is to leverage the actual CLI directly for another handful of unit tests before this is fully ready.  However, for the purposes of this PR this code is ready for merge.

File: gremlinapi/scenario_helpers.py [hunk 3]
```diff
@@ -24,50 +24,54 @@
```
Comment (khultman) [inline gremlinapi/scenario_helpers.py:24]: TODO: Evaluate deprecating this interface. IIRC the API no longs take the output this helper generates.

File: gremlinapi/__init__.py [hunk 4]
```diff
@@ -150,17 +150,17 @@ def filter(self, record):
...
-GremlinAPIConfig.user_mfa_token_value = _api_user_mfa_token
-GremlinAPIConfig.base_uri = f"{_api_host}/{_api_version}"
-GremlinAPIConfig.company_name = _api_company
-GremlinAPIConfig.api_key = _api_key
-GremlinAPIConfig.bearer_token = _api_bearer_token
-GremlinAPIConfig.bearer_timestamp = _bearer_token_timestamp
-GremlinAPIConfig.max_bearer_interval = _max_bearer_interval
-GremlinAPIConfig.http_proxy = _http_proxy
-GremlinAPIConfig.https_proxy = _https_proxy
+GremlinAPIConfig.user = _api_user  # type: ignore
```
Comment (khultman) [inline gremlinapi/__init__.py:153]: I think all of these pre-configuration variables should be strings or ints.

File: gremlinapi/templates.py [hunk 5]
```diff
@@ -13,8 +13,14 @@
     HTTPError,
 )
```
Comment (khultman) [inline gremlinapi/templates.py:15]: TODO: Evaluate deprecating this interface: IIRC this has now `officially` been deprecated from the API.

File: gremlinapi/scenario_helpers.py (hunk 3 above)
Comment (booshington) [reply gremlinapi/scenario_helpers.py:24]: Will depreciate entire scenario_helpers.py file

File: gremlinapi/templates.py (hunk 5 above)
Comment (booshington) [reply gremlinapi/templates.py:15]: Will deprecate the entire templates.py file

File: gremlinapi/__init__.py (hunk 4 above)
Comment (booshington) [reply gremlinapi/__init__.py:153]: Will remediate, in scope for this PR

File: gremlinapi/http_clients.py [hunk 6]
```diff
@@ -60,48 +67,45 @@ def header(cls, *args, **kwargs):
...
-        error_message = f"This function is not implemented, please consume the proper http library for your environment"
-        log.fatal(error_message)
+    def proxies(cls) -> dict:
+        error_message: str = f"This function is not implemented, please consume the proper http library for your environment"
+        log.error(error_message)
         raise NotImplementedError(error_message)
 
 
-class GremlineAPIRequestsClient(GremlinAPIHttpClient):
+class GremlinAPIRequestsClient(GremlinAPIHttpClient):
```
Comment (booshington) [inline gremlinapi/http_clients.py:83]: Refactor to replace Union[object, object] in all files that consume these classes

File: gremlinapi/http_clients.py [hunk 7]
```diff
@@ -148,37 +154,44 @@ def api_call(cls, method, endpoint, *args, **kwargs):
 class GremlinAPIurllibClient(GremlinAPIHttpClient):
```
Comment (booshington) [inline gremlinapi/http_clients.py:154]: Refactor to replace Union[object, object] in all files that consume these classes

**Comment by booshington** [comment]
@khultman Added missing annotations and changed the `Union[GremlinAPIRequestsClient, GremlinAPIurllibClient]` to a single `Type[GremlinAPIHttpClient]`.  In `GremlinAPIHttpClient`, the logic in the `proxy()` and `api_call()` methods have been changed to utilize the proper library given the operating conditions. When you call `GremlinAPIHttpClient.api_call()` the existence of `requests` is checked and, if so, returns `GremlinAPIRequestsClient.api_call()`. otherwise, `GremlinAPIurllibClient.api_call()` is returned.  Similarly, when `GremlinAPIHttpClient.proxies()` is called the same library check takes place. If `requests` is available `GremlinAPIRequestsClient.proxies()` is called, and if not `GremlinAPIurllibClient.proxies()` is called which returns the original intended error.

**Review by khultman — APPROVED** [review]

**Pull request merged**